mcproto/utils/abc.py

```diff
@@ -64,17 +75,58 @@ def __new__(cls: type[Self], *a, **kw) -> Self:
 
 
 class Serializable(ABC):
-    """Base class for any type that should be (de)serializable into/from :class:`~mcproto.Buffer` data."""
+    """Base class for any type that should be (de)serializable into/from :class:`~mcproto.Buffer` data.
+
+    Any class that inherits from this class and adds parameters should use the :func:`~mcproto.utils.abc.dataclass`
+    decorator.
+    """
 
     __slots__ = ()
 
-    @abstractmethod
+    def __post_init__(self) -> None:
+        """Run the validation method after the object is initialized."""
+        self.validate()
+        self.transform()
+
     def serialize(self) -> Buffer:
         """Represent the object as a :class:`~mcproto.Buffer` (transmittable sequence of bytes)."""
+        buf = Buffer()
+        self.serialize_to(buf)
+        return buf
+
+    @abstractmethod
+    def serialize_to(self, buf: Buffer, /) -> None:
+        """Write the object to a :class:`~mcproto.Buffer`."""
         raise NotImplementedError
 
+    def transform(self) -> None:
+        """Apply a transformation to the payload of the object.
+
+        This can be used to convert an int to an enum. This method is called during the initialization,
+        after the validation method. By default, this method does nothing. Override it in your subclass
+        to add transformation logic.
+
+        This method should not raise any exception.
```

Comment:
Why is that? What if the enum conversion fails, because someone passed a value like 5 when the enum only has 2 variants?